File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/ThreadSlideController.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.ui.controller;

import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.dp;
import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.inflate;

import android.content.Context;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.github.k1rakishou.ChanSettings;
import com.github.k1rakishou.chan.R;
import com.github.k1rakishou.chan.controller.Controller;
import com.github.k1rakishou.chan.controller.transition.ControllerTransition;
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent;
import com.github.k1rakishou.chan.features.drawer.MainControllerCallbacks;
import com.github.k1rakishou.chan.ui.controller.navigation.DoubleNavigationController;
import com.github.k1rakishou.chan.ui.layout.ThreadSlidingPaneLayout;
import com.github.k1rakishou.chan.ui.toolbar.NavigationItem;
import com.github.k1rakishou.chan.ui.toolbar.Toolbar;
import com.github.k1rakishou.chan.ui.widget.SlidingPaneLayoutEx;
import com.github.k1rakishou.core_themes.ThemeEngine;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;

public class ThreadSlideController
        extends Controller
        implements DoubleNavigationController,
        SlidingPaneLayoutEx.PanelSlideListener,
        ThemeEngine.ThemeChangesListener {
    private static final String TAG = "ThreadSlideController";

    @Inject
    ThemeEngine themeEngine;

    public BrowseController leftController;
    public ViewThreadController rightController;

    @Nullable
    private MainControllerCallbacks mainControllerCallbacks;
    private boolean leftOpen = true;
    private ViewGroup emptyView;
    private ThreadSlidingPaneLayout slidingPaneLayout;

    @Override
    protected void injectDependencies(@NotNull ActivityComponent component) {
        component.inject(this);
    }

    public ThreadSlideController(Context context, ViewGroup emptyView, @NotNull MainControllerCallbacks mainControllerCallbacks) {
        super(context);

        this.emptyView = emptyView;
        this.mainControllerCallbacks = mainControllerCallbacks;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        doubleNavigationController = this;

        navigation.swipeable = false;
        navigation.hasDrawer = true;

        view = inflate(context, R.layout.controller_thread_slide);

        slidingPaneLayout = view.findViewById(R.id.sliding_pane_layout);
        slidingPaneLayout.setThreadSlideController(this);
        slidingPaneLayout.setPanelSlideListener(this);
        slidingPaneLayout.setParallaxDistance(dp(100));
        slidingPaneLayout.allowedToSlide(ChanSettings.viewThreadControllerSwipeable.get());

        if (ChanSettings.isSlideLayoutMode()) {
            slidingPaneLayout.setShadowResourceLeft(R.drawable.panel_shadow);
        }

        slidingPaneLayout.openPane();

        setLeftController(null, false);
        setRightController(null, false);

        TextView textView = emptyView.findViewById(R.id.select_thread_text);
        if (textView != null) {
            textView.setTextColor(themeEngine.getChanTheme().getTextColorSecondary());
        }

        themeEngine.addListener(this);
        onThemeChanged();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        themeEngine.removeListener(this);
        mainControllerCallbacks = null;
    }

    @Override
    public void onThemeChanged() {
        if (slidingPaneLayout != null) {
            int fadeColor = (themeEngine.getChanTheme().getBackColor() & 0xffffff) + 0xCC000000;
            slidingPaneLayout.setSliderFadeColor(fadeColor);
            slidingPaneLayout.requestLayout();
        }
    }

    @Override
    public void onShow() {
        super.onShow();

        if (mainControllerCallbacks != null) {
            mainControllerCallbacks.resetBottomNavViewCheckState();
        }
    }

    public void onSlidingPaneLayoutStateRestored() {
        boolean restoredOpen = slidingPaneLayout.getPreservedOpenState();

        if (restoredOpen != leftOpen) {
            leftOpen = restoredOpen;
            slideStateChanged(false);
        }
    }

    @Override
    public void onPanelSlide(@NonNull View panel, float slideOffset) {
    }

    @Override
    public void onPanelOpened(@NonNull View panel) {
        if (this.leftOpen != leftOpen()) {
            this.leftOpen = leftOpen();
            slideStateChanged();
        }
    }

    @Override
    public void onPanelClosed(@NonNull View panel) {
        if (this.leftOpen != leftOpen()) {
            this.leftOpen = leftOpen();
            slideStateChanged();
        }
    }

    @Override
    public void switchToController(boolean leftController) {
        switchToController(leftController, true);
    }

    @Override
    public void switchToController(boolean leftController, boolean animated) {
        if (leftController != leftOpen()) {
            if (leftController) {
                slidingPaneLayout.openPane();
            } else {
                slidingPaneLayout.closePane();
            }

            requireNavController().requireToolbar().processScrollCollapse(
                    Toolbar.TOOLBAR_COLLAPSE_SHOW,
                    true
            );

            leftOpen = leftController;
            slideStateChanged(animated);
        }
    }

    public void setLeftController(@Nullable Controller leftController, boolean animated) {
        if (this.leftController != null) {
            this.leftController.onHide();
            removeChildController(this.leftController);
        }

        this.leftController = (BrowseController) leftController;

        if (leftController != null) {
            addChildController(leftController);
            leftController.attachToParentView(slidingPaneLayout.leftPane);
            leftController.onShow();
            if (leftOpen()) {
                setParentNavigationItem(true, animated);
            }
        }
    }

    public void setRightController(@Nullable Controller rightController, boolean animated) {
        if (this.rightController != null) {
            this.rightController.onHide();
            removeChildController(this.rightController);
        } else {
            this.slidingPaneLayout.rightPane.removeAllViews();
        }

        this.rightController = (ViewThreadController) rightController;

        if (rightController != null) {
            addChildController(rightController);
            rightController.attachToParentView(slidingPaneLayout.rightPane);
            rightController.onShow();
            if (!leftOpen()) {
                setParentNavigationItem(false, animated);
            }
        } else {
            slidingPaneLayout.rightPane.addView(emptyView);
        }
    }

    @Override
    public Controller getLeftController() {
        return leftController;
    }

    @Override
    public Controller getRightController() {
        return rightController;
    }

    @Override
    public void openControllerWrappedIntoBottomNavAwareController(Controller controller) {
        requireStartActivity().openControllerWrappedIntoBottomNavAwareController(controller);
    }

    @Override
    public boolean pushController(Controller to) {
        return navigationController.pushController(to);
    }

    @Override
    public boolean pushController(Controller to, boolean animated) {
        return navigationController.pushController(to, animated);
    }

    @Override
    public boolean pushController(Controller to, ControllerTransition controllerTransition) {
        return navigationController.pushController(to, controllerTransition);
    }

    @Override
    public boolean popController() {
        return navigationController.popController();
    }

    @Override
    public boolean popController(boolean animated) {
        return navigationController.popController(animated);
    }

    @Override
    public boolean popController(ControllerTransition controllerTransition) {
        return navigationController.popController(controllerTransition);
    }

    @Override
    public boolean onBack() {
        if (!leftOpen()) {
            if (rightController != null && rightController.onBack()) {
                return true;
            } else {
                switchToController(true);
                return true;
            }
        } else {
            if (leftController != null && leftController.onBack()) {
                return true;
            }
        }

        return super.onBack();
    }

    public boolean leftOpen() {
        return slidingPaneLayout.isOpen();
    }

    private void slideStateChanged() {
        slideStateChanged(true);
    }

    private void slideStateChanged(boolean animated) {
        setParentNavigationItem(leftOpen, animated);

        if (leftOpen && rightController != null) {
            ((ReplyAutoCloseListener) rightController).onReplyViewShouldClose();
        } else if (!leftOpen && leftController != null) {
            ((ReplyAutoCloseListener) leftController).onReplyViewShouldClose();
        }

        notifyFocusLost(
                leftOpen ? ThreadControllerType.Thread : ThreadControllerType.Catalog,
                leftOpen ? rightController : leftController
        );

        notifyFocusGained(
                leftOpen ? ThreadControllerType.Catalog : ThreadControllerType.Thread,
                leftOpen ? leftController : rightController
        );
    }

    private void notifyFocusLost(ThreadControllerType controllerType, Controller controller) {
        if (controller == null) {
            return;
        }

        if (controller instanceof SlideChangeListener) {
            ((SlideChangeListener) controller).onLostFocus(controllerType);
        }

        for (Controller childController : controller.childControllers) {
            notifyFocusGained(controllerType, childController);
        }
    }

    private void notifyFocusGained(ThreadControllerType controllerType, Controller controller) {
        if (controller == null) {
            return;
        }

        if (controller instanceof SlideChangeListener) {
            ((SlideChangeListener) controller).onGainedFocus(controllerType);
        }

        for (Controller childController : controller.childControllers) {
            notifyFocusGained(controllerType, childController);
        }
    }

    private void setParentNavigationItem(boolean left, boolean animate) {
        Toolbar toolbar = requireNavController().requireToolbar();

        // default, blank navigation item with no menus or titles, so other layouts don't mess up
        NavigationItem item = new NavigationItem();
        if (left) {
            if (leftController != null) {
                item = leftController.navigation;
            }
        } else {
            if (rightController != null) {
                item = rightController.navigation;
            }
        }

        navigation = item;
        navigation.swipeable = false;
        navigation.hasDrawer = true;

        toolbar.setNavigationItem(animate, true, navigation, null);
    }

    public boolean passMotionEventIntoSlidingPaneLayout(@NotNull MotionEvent event) {
        return slidingPaneLayout.onTouchEvent(event);
    }

    public interface ReplyAutoCloseListener {
        void onReplyViewShouldClose();
    }

    public interface SlideChangeListener {
        void onGainedFocus(@NonNull ThreadControllerType controllerType);
        void onLostFocus(@NonNull ThreadControllerType controllerType);
    }

    public enum ThreadControllerType {
        Catalog,
        Thread
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/ThreadSlideController.kt
package com.github.k1rakishou.chan.ui.controller

import android.content.Context
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.isSpecified
import androidx.compose.ui.graphics.toArgb
import com.github.k1rakishou.chan.R
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent
import com.github.k1rakishou.chan.features.drawer.MainControllerCallbacks
import com.github.k1rakishou.chan.features.toolbar.BackArrowMenuItem
import com.github.k1rakishou.chan.features.toolbar.HamburgMenuItem
import com.github.k1rakishou.chan.features.toolbar.KurobaToolbarState
import com.github.k1rakishou.chan.ui.controller.base.Controller
import com.github.k1rakishou.chan.ui.controller.base.DeprecatedNavigationFlags
import com.github.k1rakishou.chan.ui.controller.base.transition.ControllerTransition
import com.github.k1rakishou.chan.ui.controller.base.transition.TransitionMode
import com.github.k1rakishou.chan.ui.controller.navigation.DoubleNavigationController
import com.github.k1rakishou.chan.ui.globalstate.reply.ReplyLayoutBoundsStates
import com.github.k1rakishou.chan.ui.layout.ThreadSlidingPaneLayout
import com.github.k1rakishou.chan.ui.view.widget.SlidingPaneLayoutEx
import com.github.k1rakishou.chan.ui.viewstate.ReplyLayoutVisibilityStates
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils
import com.github.k1rakishou.core_themes.ThemeEngine
import com.github.k1rakishou.core_themes.ThemeEngine.ThemeChangesListener
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject

class ThreadSlideController(
  context: Context,
  mainControllerCallbacks: MainControllerCallbacks,
  private val emptyView: ViewGroup
) : Controller(context),
  DoubleNavigationController,
  SlidingPaneLayoutEx.PanelSlideListener,
  ThemeChangesListener {

  @Inject
  lateinit var themeEngine: ThemeEngine

  private val _leftController = MutableStateFlow<BrowseController?>(null)
  override val leftControllerFlow: StateFlow<Controller?>
    get() = _leftController.asStateFlow()

  private val _rightController = MutableStateFlow<ViewThreadController?>(null)
  override val rightControllerFlow: StateFlow<Controller?>
    get() = _rightController.asStateFlow()

  private var mainControllerCallbacks: MainControllerCallbacks?
  private var slidingPaneLayout: ThreadSlidingPaneLayout? = null
  private var slidingPaneLayoutOpenState = SlidingPaneLayoutOpenState.LeftOpened
  private var isSlidingInProgress = false

  private val emptyCatalogToolbar by lazy(LazyThreadSafetyMode.NONE) {
    val kurobaToolbarState = KurobaToolbarState(
      controllerHash = 0, // Hardcoded
      controllerKey = controllerKey,
      globalUiStateHolder = globalUiStateHolder
    )

    kurobaToolbarState.enterDefaultMode(
      leftItem = HamburgMenuItem(
        onClick = {
          // no-op
        }
      )
    )

    return@lazy kurobaToolbarState
  }

  private val emptyThreadToolbar by lazy(LazyThreadSafetyMode.NONE) {
    val kurobaToolbarState = KurobaToolbarState(
      controllerHash = 1, // Hardcoded
      controllerKey = controllerKey,
      globalUiStateHolder = globalUiStateHolder
    )

    kurobaToolbarState.enterDefaultMode(
      leftItem = BackArrowMenuItem(
        onClick = { switchToLeftController() }
      )
    )

    return@lazy kurobaToolbarState
  }

  override fun injectActivityDependencies(component: ActivityComponent) {
    component.inject(this)
  }

  init {
    this.mainControllerCallbacks = mainControllerCallbacks
  }

  override val toolbarState: KurobaToolbarState
    get() = getToolbarState(slidingPaneLayoutOpenState)

  override val leftControllerToolbarState: KurobaToolbarState?
    get() = leftController()?.toolbarState
  override val rightControllerToolbarState: KurobaToolbarState?
    get() = rightController()?.toolbarState

  override fun onCreate() {
    super.onCreate()

    doubleNavigationController = this

    updateNavigationFlags(
      newNavigationFlags = DeprecatedNavigationFlags(
        swipeable = false,
        hasDrawer = true
      )
    )

    toolbarState.enterDefaultMode(
      leftItem = null,
      middleContent = null
    )

    view = AppModuleAndroidUtils.inflate(context, R.layout.controller_thread_slide)
    slidingPaneLayout = view.findViewById<ThreadSlidingPaneLayout>(R.id.sliding_pane_layout).also { slidingPane ->
      slidingPane.setThreadSlideController(this)
      slidingPane.setPanelSlideListener(this)
      slidingPane.setParallaxDistance(AppModuleAndroidUtils.dp(100f))
      slidingPane.allowedToSlide(kurobaSettings.application.viewThreadControllerSwipeable.readBlocking())

      if (kurobaSettings.application.isSlideLayoutModeBlocking()) {
        slidingPane.setShadowResourceLeft(R.drawable.panel_shadow)
      }

      slidingPane.openPane()
    }

    updateLeftController(newLeftController = null, animated = false)
    updateRightController(newRightController = null, animated = false)

    val textView = emptyView.findViewById<TextView>(R.id.select_thread_text)
    textView?.setTextColor(themeEngine.chanTheme.textColorSecondary)

    themeEngine.addListener(this)
    onThemeChanged()

    controllerScope.launch {
      combine(
        flow = globalUiStateHolder.replyLayout.replyLayoutVisibilityEventsFlow,
        flow2 = globalUiStateHolder.replyLayout.replyLayoutsBoundsFlow,
        flow3 = globalUiStateHolder.mainUi.touchPosition,
        transform = { replyLayoutVisibilityEvents, replyLayoutsBounds, touchPosition ->
          return@combine SlidingPaneLockState(
            replyLayoutVisibilityStates = replyLayoutVisibilityEvents,
            replyLayoutsBounds = replyLayoutsBounds,
            touchPosition = touchPosition,
          )
        }
      )
        .onEach { slidingPaneLockState -> slidingPaneLayout?.lockUnlockSliding(slidingPaneLockState.isLocked()) }
        .collect()
    }
  }

  override fun onDestroy() {
    super.onDestroy()
    themeEngine.removeListener(this)
    mainControllerCallbacks = null
  }

  override fun onThemeChanged() {
    slidingPaneLayout?.sliderFadeColor = themeEngine.chanTheme.backColorCompose.copy(alpha = 0.7f).toArgb()
    slidingPaneLayout?.coveredFadeColor = androidx.compose.ui.graphics.Color.Black.copy(alpha = 0.5f).toArgb()
  }

  fun onSlidingPaneLayoutStateRestored() {
    val restoredOpen = slidingPaneLayout?.preservedOpenState
      ?: return

    val newSlidingPaneLayoutOpenState = SlidingPaneLayoutOpenState.from(restoredOpen)
    if (newSlidingPaneLayoutOpenState != slidingPaneLayoutOpenState) {
      slidingPaneLayoutOpenState = newSlidingPaneLayoutOpenState
      slideStateChanged(false)
    }
  }

  override fun onPanelSlide(panel: View, slideOffset: Float) {
    if (!isSlidingInProgress) {
      isSlidingInProgress = true
      startToolbarTransition()
    } else {
      updateToolbarTransition(slideOffset)
    }
  }

  override fun onPanelOpened(panel: View) {
    if (slidingPaneLayoutOpenState != SlidingPaneLayoutOpenState.from(isLeftOpen())) {
      slidingPaneLayoutOpenState = SlidingPaneLayoutOpenState.LeftOpened
      slideStateChanged()
    }

    finishToolbarTransition(slidingPaneLayoutOpenState)
  }

  override fun onPanelClosed(panel: View) {
    if (slidingPaneLayoutOpenState != SlidingPaneLayoutOpenState.from(isLeftOpen())) {
      slidingPaneLayoutOpenState = SlidingPaneLayoutOpenState.RightOpened
      slideStateChanged()
    }

    finishToolbarTransition(slidingPaneLayoutOpenState)
  }

  override fun switchToLeftController(animated: Boolean) {
    switchToControllerInternal(left = true, animated = animated)
  }

  override fun switchToRightController(animated: Boolean) {
    switchToControllerInternal(left = false, animated = animated)
  }

  override fun pushToLeftController(controller: Controller, animated: Boolean) {
    TODO("Not yet implemented (pushToLeftController)")
  }

  override fun pushToRightController(controller: Controller, animated: Boolean) {
    TODO("Not yet implemented (pushToRightController)")
  }

  override fun updateLeftController(newLeftController: Controller?, animated: Boolean) {
    val currentLeftController = leftController()
    if (currentLeftController != null) {
      currentLeftController.onHide()
      removeChildController(currentLeftController)
    }

    _leftController.value = newLeftController as BrowseController?

    if (newLeftController != null && slidingPaneLayout != null) {
      addChildController(newLeftController)
      newLeftController.attachToParentView(slidingPaneLayout!!.leftPane)
      newLeftController.onShow()

      if (isLeftOpen()) {
        updateContainerToolbarStateWithChildToolbarState(
          slidingPaneLayoutOpenState = SlidingPaneLayoutOpenState.LeftOpened,
          animate = animated
        )
      }
    }
  }

  override fun updateRightController(newRightController: Controller?, animated: Boolean) {
    val currentRightController = rightController()
    if (currentRightController != null) {
      currentRightController.onHide()
      removeChildController(currentRightController)
    } else {
      slidingPaneLayout?.rightPane?.removeAllViews()
    }

    _rightController.value = newRightController as ViewThreadController?

    if (newRightController != null) {
      if (slidingPaneLayout != null) {
        addChildController(newRightController)
        newRightController.attachToParentView(slidingPaneLayout!!.rightPane)
        newRightController.onShow()

        if (!isLeftOpen()) {
          updateContainerToolbarStateWithChildToolbarState(
            slidingPaneLayoutOpenState = SlidingPaneLayoutOpenState.RightOpened,
            animate = animated
          )
        }
      }
    } else {
      slidingPaneLayout?.rightPane?.addView(emptyView)
    }
  }

  override fun leftController(): BrowseController? {
    return _leftController.value
  }

  override fun rightController(): ViewThreadController? {
    return _rightController.value
  }

  override fun pushController(to: Controller): Boolean {
    return navigationController?.pushController(to) ?: false
  }

  override fun pushController(to: Controller, animated: Boolean): Boolean {
    return navigationController?.pushController(to, animated) ?: false
  }

  override fun pushController(to: Controller, onFinished: () -> Unit): Boolean {
    return navigationController?.pushController(to, onFinished) ?: false
  }

  override fun pushController(to: Controller, transition: ControllerTransition?): Boolean {
    return navigationController?.pushController(to, transition) ?: false
  }

  override fun popController(): Boolean {
    return navigationController?.popController() ?: false
  }

  override fun popController(animated: Boolean): Boolean {
    return navigationController?.popController(animated) ?: false
  }

  override fun popController(onFinished: () -> Unit): Boolean {
    return navigationController?.popController(onFinished) ?: false
  }

  override fun popController(transition: ControllerTransition?): Boolean {
    return navigationController?.popController(transition) ?: false
  }

  override fun onBack(): Boolean {
    if (slidingPaneLayout != null) {
      if (isRightOpen()) {
        if (rightController()?.onBack() == true) {
          return true
        }

        switchToLeftController()
        return true
      }

      if (leftController()?.onBack() == true) {
        return true
      }
    }

    return super.onBack()
  }

  fun isLeftOpen(): Boolean {
    return slidingPaneLayout!!.isOpen
  }

  fun isRightOpen(): Boolean {
    return !isLeftOpen()
  }

  private fun switchToControllerInternal(left: Boolean, animated: Boolean) {
    if (left != isLeftOpen()) {
      if (left) {
        slidingPaneLayout?.openPane()
      } else {
        slidingPaneLayout?.closePane()
      }

      toolbarState.showToolbar()
    }
  }

  private fun slideStateChanged(animated: Boolean = true) {
    updateContainerToolbarStateWithChildToolbarState(
      slidingPaneLayoutOpenState = slidingPaneLayoutOpenState,
      animate = animated
    )

    val rightController = rightController()
    val leftController = leftController()

    if (slidingPaneLayoutOpenState.leftOpenedOrOpening && rightController != null) {
      (rightController as ReplyAutoCloseListener).onReplyViewShouldClose()
    } else if (slidingPaneLayoutOpenState.rightOpenedOrOpening && leftController != null) {
      (leftController as ReplyAutoCloseListener).onReplyViewShouldClose()
    }

    notifyFocusLost(
      controllerType = if (slidingPaneLayoutOpenState.leftOpenedOrOpening) {
        ThreadControllerType.Thread
      } else {
        ThreadControllerType.Catalog
      },
      controller = if (slidingPaneLayoutOpenState.leftOpenedOrOpening) {
        rightController
      } else {
        leftController
      }
    )

    notifyFocusGained(
      controllerType = if (slidingPaneLayoutOpenState.leftOpenedOrOpening) {
        ThreadControllerType.Catalog
      } else {
        ThreadControllerType.Thread
      },
      controller = if (slidingPaneLayoutOpenState.leftOpenedOrOpening) {
        leftController
      } else {
        rightController
      }
    )
  }

  private fun notifyFocusLost(controllerType: ThreadControllerType, controller: Controller?) {
    if (controller == null) {
      return
    }

    if (controller is SlideChangeListener) {
      (controller as SlideChangeListener).onLostFocus(controllerType)
    }

    for (childController in controller.childControllers) {
      notifyFocusLost(controllerType, childController)
    }
  }

  private fun notifyFocusGained(controllerType: ThreadControllerType, controller: Controller?) {
    if (controller == null) {
      return
    }

    if (controller is SlideChangeListener) {
      (controller as SlideChangeListener).onGainedFocus(controllerType)
    }

    for (childController in controller.childControllers) {
      notifyFocusGained(controllerType, childController)
    }
  }

  private fun updateContainerToolbarStateWithChildToolbarState(
    slidingPaneLayoutOpenState: SlidingPaneLayoutOpenState,
    animate: Boolean
  ) {
    containerToolbarState = getToolbarState(slidingPaneLayoutOpenState)
  }

  private fun getToolbarState(slidingPaneLayoutOpenState: SlidingPaneLayoutOpenState): KurobaToolbarState {
    var kurobaToolbarState = if (slidingPaneLayoutOpenState.leftOpenedOrOpening) {
      leftController()?.toolbarState
    } else {
      rightController()?.toolbarState
    }

    if (kurobaToolbarState == null) {
      kurobaToolbarState = if (slidingPaneLayoutOpenState.leftOpenedOrOpening) {
        emptyCatalogToolbar
      } else {
        emptyThreadToolbar
      }
    }

    return kurobaToolbarState
  }

  private fun startToolbarTransition() {
    val transitionMode = if (slidingPaneLayoutOpenState.rightOpenedOrOpening) {
      TransitionMode.In
    } else {
      TransitionMode.Out
    }

    val kurobaToolbarState = getToolbarState(slidingPaneLayoutOpenState.invert())

    containerToolbarState.onTransitionProgressStart(
      other = kurobaToolbarState,
      transitionMode = transitionMode
    )
  }

  private fun updateToolbarTransition(slideOffset: Float) {
    containerToolbarState.onTransitionProgress(
      progress = slideOffset
    )
  }

  private fun finishToolbarTransition(slidingPaneLayoutOpenState: SlidingPaneLayoutOpenState) {
    if (!isSlidingInProgress) {
      return
    }

    val prevToolbarState = containerToolbarState
    containerToolbarState = getToolbarState(slidingPaneLayoutOpenState)
    prevToolbarState.onTransitionProgressFinished()

    isSlidingInProgress = false
  }

  fun passMotionEventIntoSlidingPaneLayout(event: MotionEvent): Boolean {
    return slidingPaneLayout?.onTouchEvent(event) ?: false
  }

  interface ReplyAutoCloseListener {
    fun onReplyViewShouldClose()
  }

  interface SlideChangeListener {
    fun onGainedFocus(wasFocused: ThreadControllerType)
    fun onLostFocus(nowFocused: ThreadControllerType)
  }

  private data class SlidingPaneLockState(
    val replyLayoutVisibilityStates: ReplyLayoutVisibilityStates,
    val replyLayoutsBounds: ReplyLayoutBoundsStates,
    val touchPosition: Offset,
  ) {

    fun isLocked(): Boolean {
      if (replyLayoutVisibilityStates.anyExpanded()) {
        return true
      }

      if (touchPosition.isSpecified) {
        if (!replyLayoutsBounds.catalog.isEmpty && !replyLayoutVisibilityStates.catalog.isCollapsed()) {
          if (replyLayoutsBounds.catalog.contains(touchPosition)) {
            return true
          }
        }

        if (!replyLayoutsBounds.thread.isEmpty && !replyLayoutVisibilityStates.thread.isCollapsed()) {
          if (replyLayoutsBounds.thread.contains(touchPosition)) {
            return true
          }
        }
      }

      return false
    }

  }

  private enum class SlidingPaneLayoutOpenState {
    LeftOpening,
    LeftOpened,
    RightOpening,
    RightOpened;

    fun asTransition(): SlidingPaneLayoutOpenState {
      return when (this) {
        LeftOpened -> LeftOpening
        RightOpened -> RightOpening
        LeftOpening -> error("Shouldn't happen")
        RightOpening -> error("Shouldn't happen")
      }
    }

    fun invert(): SlidingPaneLayoutOpenState {
      return when (this) {
        LeftOpening -> RightOpening
        LeftOpened -> RightOpened
        RightOpening -> LeftOpening
        RightOpened -> LeftOpened
      }
    }

    val leftOpenedOrOpening: Boolean
      get() = this == LeftOpened || this == LeftOpening
    val rightOpenedOrOpening: Boolean
      get() = this == RightOpened || this == RightOpening

    companion object {
      fun from(leftOpened: Boolean): SlidingPaneLayoutOpenState {
        return if (leftOpened) {
          LeftOpened
        } else {
          RightOpened
        }
      }
    }
  }

  companion object {
    private const val TAG = "ThreadSlideController"
  }
}
